dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/impl/AggregateContext.kt

```diff
@@ -0,0 +1,135 @@
+import it.unibo.collektive.aggregate.api.operators.neighbouring
+import it.unibo.collektive.field.Field
+import it.unibo.collektive.networking.InboundMessage
+import it.unibo.collektive.networking.OutboundMessage
+import it.unibo.collektive.networking.SingleOutboundMessage
+import it.unibo.collektive.path.Path
+import it.unibo.collektive.state.State
+import it.unibo.collektive.state.impl.getTyped
+
+/**
+ * Context for managing aggregate computation.
+ * It represents the [localId] of the device, the [messages] received from the neighbours,
+ * and the [previousState] of the device.
+ */
+internal class AggregateContext(
+    override val localId: ID,
+    private val messages: Iterable<InboundMessage>,
+    private val previousState: State,
+) : Aggregate {
+
+    private val stack = Stack()
+    private var state: State = mapOf()
+    private var toBeSent = OutboundMessage(localId, emptyMap())
+
+    /**
+     * Messages to send to the other nodes.
+     */
+    fun messagesToSend(): OutboundMessage = toBeSent
+
+    /**
+     * Return the current state of the device as a new state.
+     */
+    fun newState(): State = state
+
+    private fun <T> newField(localValue: T, others: Map<ID, T>): Field<T> = Field(localId, localValue, others)
+
+    override fun <X> exchange(initial: X, body: (Field<X>) -> Field<X>): Field<X> {
+        val messages = messagesAt<X>(stack.currentPath())
+        val previous = stateAt(stack.currentPath(), initial)
+        val subject = newField(previous, messages)
+        return body(subject).also { field ->
+            val message = SingleOutboundMessage(field.localValue, field.excludeSelf())
+            val path = stack.currentPath()
+            check(!toBeSent.messages.containsKey(path)) {
+                """
+                    Alignment was broken by multiple aligned calls with the same path: $path.
+                    The most likely cause is an aggregate function call within a loop
+                """.trimIndent()
+            }
+            toBeSent = toBeSent.copy(messages = toBeSent.messages + (stack.currentPath() to message))
+            state = state + (stack.currentPath() to field.localValue)
+        }
+    }
+
+    override fun <Init, Ret> exchanging(
+        initial: Init,
+        body: YieldingContext<Field<Init>, Field<Ret>>.(Field<Init>) -> YieldingResult<Field<Init>, Field<Ret>>,
+    ): Field<Ret> {
+        val messages = messagesAt<Init>(stack.currentPath())
+        val previous = stateAt(stack.currentPath(), initial)
+        val subject = newField(previous, messages)
+        val context = YieldingContext<Field<Init>, Field<Ret>>()
+        var res: Option<YieldingResult<Field<Init>, Field<Ret>>>
+        body(context, subject).also {
+            res = it.some()
+            state = state + (stack.currentPath() to it.toReturn.localValue)
+        }
+        return res.getOrElse { error("This error should never be thrown") }.toReturn
+    }
```

Comment:
Why do we need to use optionals here? Seems a complete overkill

```suggestion
        return body(context, subject).also {
            state = state + (stack.currentPath() to it.toReturn.localValue)
        }.toReturn
    }
```

Also why are we saving `toReturn`? I think it's bugged, we should be saving `toSend`, shouldn't we?

Reply:
Yes, there was a bug in the actual implementation of the exchanging and also the exchange.
A fix is incoming.

Reply:
Before fixing it, please add a test that triggers the disaster :smile:

Reply:
Pushed a minimal test checking both `toSend` and the return of the `exchanging`